torch_geometric/data/fs_utils.py

```diff
@@ -0,0 +1,137 @@
+import io
+import os.path as osp
+import sys
+from typing import Any, Optional
+
+import fsspec
+import torch
+from fsspec.core import url_to_fs
+
+DEFAULT_CACHE_PATH = '/tmp/pyg_simplecache'
+
+
+def get_fs(path: str) -> fsspec.AbstractFileSystem:
+    """Get filesystem backend given a path URI to the resource.
+
+    Here are some common example paths and dispatch result:
+    * /home/file -> fsspec.implementations.local.LocalFileSystem.
+    * memory://home/file -> fsspec.implementations.memory.MemoryFileSystem.
+    * https://home/file -> fsspec.implementations.http.HTTPFileSystem.
+    * gs://home/file -> gcsfs.GCSFileSystem.
+    * s3://home/file -> s3fs.S3FileSystem.
+
+    See current list backend implementations in fsspec, here:
+    https://github.com/fsspec/filesystem_spec/blob/master/fsspec/registry.py#L62.
+
+    The backend dispatch can be updated with custom backends following:
+    https://filesystem-spec.readthedocs.io/en/latest/developer.html#implementing-a-backend.
+
+    Args:
+        path: URI to filesystem location, e.g., gs://home/me/file, s3://...
+
+    Returns:
+        fsspec filesystem backend for the URI.
+    """
+    return url_to_fs(path)[0]
+
+
+def fs_normpath(path: str, *args) -> str:
+    if get_fs(path).protocol == 'file':
+        return osp.normpath(path, *args)
+    return path
+
+
+def fs_exists(path: str) -> bool:
+    return get_fs(path).exists(path)
+
+
+def fs_ls(path: str, detail: bool = False) -> bool:
+    fs = get_fs(path)
+    results = fs.ls(path, detail=detail)
+    # TODO: Strip common ancestor.
+    return [osp.basename(x) for x in results]
+
+
+def fs_mkdirs(path: str, **kwargs):
+    return get_fs(path).mkdirs(path, **kwargs)
+
+
+def fs_isdir(path: str) -> bool:
+    return get_fs(path).isdir(path)
+
+
+def _copy_file_handle(f_from: Any, path: str, blocksize: int, **kwargs):
+    with fsspec.open(path, 'wb', **kwargs) as f_to:
+        data = True
+        while data:
+            data = f_from.read(blocksize)
+            f_to.write(data)
+
+
+def fs_cp(path1: str, path2: str, kwargs1: Optional[dict] = None,
+          kwargs2: Optional[dict] = None, extract: bool = False,
+          cache_path: Optional[str] = DEFAULT_CACHE_PATH,
+          blocksize: int = 2 * 22):
+    # Initialize kwargs for source/destination.
+    kwargs1 = kwargs1 or {}
+    kwargs2 = kwargs2 or {}
+
+    # Cache result if the protocol is not local and we have a cache folder.
+    if get_fs(path1).protocol != 'file' and cache_path:
```

Comment:
@rusty1s I'm pointing out 3 things that fs_cp is doing that might be surprising to help you focus on this function. Here we automatically add a simplecache stage to the path so that results of a remote source path will first go to a local filesystem cache location so that subsequent fetches will succeed without making remote requests, e.g., http, gs, s3, etc.